common/interpolation/include/interpolation/interpolation_utils.hpp

```diff
@@ -15,48 +15,78 @@
 #ifndef INTERPOLATION__INTERPOLATION_UTILS_HPP_
 #define INTERPOLATION__INTERPOLATION_UTILS_HPP_
 
+#include <algorithm>
 #include <array>
+#include <functional>
+#include <limits>
 #include <stdexcept>
 #include <vector>
 
 namespace interpolation_utils
 {
+constexpr double EPS = std::numeric_limits<double>::epsilon();
+
+template <class T, typename std::enable_if_t<std::is_floating_point_v<T>, bool> * = nullptr>
+bool isEqual(T a, T b)
+{
+  return abs(a - b) < std::numeric_limits<T>::epsilon();
+}
+
+// Strictly monotonic increasing
 inline bool isIncreasing(const std::vector<double> & x)
 {
   if (x.empty()) {
-    throw std::invalid_argument("Points is empty.");
+    throw std::invalid_argument("The vector x is empty.");
   }
 
-  for (size_t i = 0; i < x.size() - 1; ++i) {
-    if (x.at(i) >= x.at(i + 1)) {
-      return false;
-    }
-  }
+  // x(i) <= x(i+1) condition check
+  auto const & is_increasing_it = std::adjacent_find(x.begin(), x.end(), std::greater_equal<>());
 
-  return true;
+  return is_increasing_it == x.cend();
 }
 
+// Strictly increasing and monotonic
 inline bool isNotDecreasing(const std::vector<double> & x)
 {
   if (x.empty()) {
-    throw std::invalid_argument("Points is empty.");
+    throw std::invalid_argument("The vector x is empty.");
   }
 
-  for (size_t i = 0; i < x.size() - 1; ++i) {
-    if (x.at(i) > x.at(i + 1)) {
-      return false;
-    }
+  // x(i) < x(i+1) condition check ( x(i) == x(i+1) is allowed)
+  auto const & is_strictly_increasing_it = std::adjacent_find(x.begin(), x.end(), std::greater<>());
+
+  return is_strictly_increasing_it == x.cend();
+}
+
+inline bool isStrictlyMonotonic(const std::vector<double> & x)
```

Comment:
Could you tell me why this condition is required please.

Reply:
Before this modification, there were only monotonically increasing checks. However, it does not matter whether the base coordinate is increasing or decreasing. This method is the mathematical description of monotonicity. Sometimes we might need to interpolate decreasing monotonic coordinates (0: -1, -2, ... -5, for example), sometimes increasing.

Reply:
Currently, the key is `s` of the path, but as you said, we can use decreasing values. I got it.